qiskit/transpiler/preset_passmanagers/level1.py

```diff
@@ -104,14 +107,66 @@ def level_1_pass_manager(pass_manager_config: PassManagerConfig) -> PassManager:
     # 1. Use trivial layout if no layout given
     _given_layout = SetLayout(initial_layout)
 
-    _choose_layout_and_score = [
-        TrivialLayout(coupling_map),
-        Layout2qDistance(coupling_map, property_name="trivial_layout_score"),
-    ]
-
     def _choose_layout_condition(property_set):
         return not property_set["layout"]
 
+    def _trivial_not_perfect(property_set):
+        # Verify that a trivial layout is perfect. If trivial_layout_score > 0
+        # the layout is not perfect. The layout is unconditionally set by trivial
+        # layout so we need to clear it before contuing.
+        if property_set["trivial_layout_score"] is not None:
+            if property_set["trivial_layout_score"] != 0:
+                return True
+            else:
+                warnings.warn(
+                    "The current implicit default of using a trivial layout if it's a "
+                    "perfect layout will change in a future release, if you're "
+                    "depending on this behavior it is better to explicit set "
+                    "layout_method='trivial' when calling transpile()",
+                    FutureWarning,
+                )
+                # stack_level is not set because targeting transpile() is difficult
+        return False
+
+    def _vf2_match_not_found(property_set):
+        # If a layout hasn't been set by the time we run vf2 layout we need to
+        # run layout
+        if property_set["layout"] is None:
+            return True
+        # if VF2 layout stopped for any reason other than solution found we need
+        # to run layout since VF2 didn't converge.
+        if (
+            property_set["VF2Layout_stop_reason"] is not None
+            and property_set["VF2Layout_stop_reason"].value != "solution found"
+        ):
+            return True
+        return False
+
+    vf2_seed = seed_transpiler
+    if vf2_seed is None:
+        vf2_seed = -1
+
+    _choose_layout_0 = (
+        []
+        if pass_manager_config.layout_method
+        else [
+            TrivialLayout(coupling_map),
+            Layout2qDistance(coupling_map, property_name="trivial_layout_score"),
+        ]
+    )
+
+    _choose_layout_1 = (
+        []
+        if pass_manager_config.layout_method
+        else VF2Layout(
+            coupling_map,
+            seed=vf2_seed,
+            call_limit=int(5e5),
+            time_limit=1e-2,
```

Comment:
Should we document/comment on these magic numbers?  (etc in other files)

Reply:
Yeah, that's a good idea. The call limit is at least magic which I got from manual testing (running with `int(1e7)` locally was ~20 seconds). So I used that and assumed linear scaling with the call limit to find a time around what we want the maximum time for the pass to be in each optimization level. But we definitely want this documented in the code to explain it (and also to tweak it if performance changes in future retworkx releases)

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/7213/commits/59ef3268bda527b9c5a08cadcdae5ccb737c866b